core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -1367,9 +1368,108 @@ public AlluxioURI translateUri(String uriStr) throws InvalidPathException {
     // Scheme-less URIs are regarded as Alluxio URI.
     if (uri.getScheme() == null || uri.getScheme().equals(Constants.SCHEME)) {
       return uri;
-    } else {
-      // Reverse lookup mount table to find mount point that contains the path.
+    }
+
+    // Reverse lookup mount table to find mount point that contains the path.
+    try {
       return mMountTable.translate(uriStr);
+    } catch (InvalidPathException e) {
+      // Throw if auto-mount is not requested.
+      if (!ServerConfiguration.getBoolean(PropertyKey.MASTER_AUTO_MOUNT_ENABLED)) {
+        throw e;
+      }
+    }
+
+    // Reverse lookup failed and auto-mount is requested.
+    try {
+      autoMountUfsForUri(uri);
+    } catch (Exception e) {
+      LOG.warn(String.format("Failed to auto-mount ufs for URI: %s", uri), e);
+      if (e instanceof InvalidPathException) {
+        // Don't throw as this could be thrown due to ufs having been mounted concurrently.
+      } else {
+        throw new InvalidPathException(
+            String.format("Failed to translate and auto-mount path: %s", uriStr), e);
+      }
+    }
+    // Try translating the URI after auto-mount.
+    return mMountTable.translate(uriStr);
+  }
+
+  /**
+   * Attempts to mount given UFS Uri into designated auto-mount directory.
+   *
+   * Mount attempt will start from the mounting lowest path component, and go to higher path
+   * components if mounting fails due to access control violation.
+   *
+   * TODO(ggezer) Have fine-grained locking on mount paths instead of marking synchronized.
+   *
+   * @param ufsUri ufs Uri for auto-mount
+   * @throws AlluxioException
+   * @throws IOException
+   */
+  private synchronized void autoMountUfsForUri(AlluxioURI ufsUri)
```

Comment:
what does this `synchronized` protect? I don't think any other methods are `synchronized` in this class, so this is only used for serializing auto-mount commands?

Reply:
Correct. I have a TODO for having a fine-grained locking. 

We should be locking for each unique "scheme/authority" pair. Otherwise concurrently attempting to mount the same mount point will have problems. (if one fails due to conflict, we don't know when the other is done mounting it before we serve the request...)